Add model registries and Model.from_dict method

QRemy opened the pull request:
add spatial_models dict to gammapy.image.models
add spectral_modelcs dict to gammapy.spectrum.models
add tag to model
add from_dict() to model
modify serialization.io to use model.from_dict()

cdeil commented on gammapy/spectrum/models.py:
```diff
@@ -1543,7 +1547,8 @@ def __init__(
         self.parameter = parameter
         self.parameter_name = parameter_name
-
+        self.tag = AbsorbedSpectralModel
```
tag should be a string, no?

cdeil commented on gammapy/spectrum/models.py:
```diff
@@ -1854,3 +1862,23 @@ def evaluate(energy, norm, mean, sigma):
+            "SpectralGaussian": SpectralGaussian,
+            "SpectralLogGaussian": SpectralLogGaussian,
+            "ScaleModel": ScaleModel,
+#               TODO: add support for these models writing their .from_dict()
```
Maybe `make black` doesn't format that dict properly because of the inline comment?
Suggest to move the comment before or after the dict.

adonath commented on gammapy/image/models/__init__.py:
```diff
@@ -1,3 +1,13 @@
 """Spatial 2D image models."""
 from .core import *
+
+spatial_models = {
```
There is a convention, that module level variables should be all upper case. So just rename `spatial_models` to `SPATIAL_MODELS`.

adonath commented on gammapy/image/models/core.py:
```diff
@@ -150,7 +150,7 @@ def __init__(self, lon_0, lat_0, sigma, frame="galactic"):
         self.lat_0 = Parameter("lat_0", Latitude(lat_0), min=-90, max=90)
         self.sigma = Parameter("sigma", Angle(sigma), min=0)
-
+        self.tag = "SkyGaussian"
```
I think we should define the tags as "class level" attributes, here is an example:
```python3
class A:
    a = 1  # class level attributes 
    def __init__(self, b):
        self.b = b  # instance level attribute 
```
The difference is, that class level attributes are shared across all instances of a model, while the instance attributes (such as parameter values...) can differ form instance to instance of the same model. The tag is certainly an attribute, that is shared by all instances. So one `SkyGaussian` does not have a different tag then another `SkyGaussian`. So just move the definition of the `tag` attribute up below the definition of the class name, just as in the example I've shown.

adonath commented on gammapy/spectrum/models.py:
```diff
@@ -1854,3 +1862,23 @@ def evaluate(energy, norm, mean, sigma):
         )
+
+
+spectral_models = {
```
Again just rename `spectral_models` to `SPECTRAL_MODELS`...

cdeil commented:
@QRemy - I see two test fails: https://gist.github.com/cdeil/f7058f47fb4417f73f4feb24f5a40616

QRemy commented:
> @QRemy - I see two test fails: https://gist.github.com/cdeil/f7058f47fb4417f73f4feb24f5a40616

I removed the previous .from_dict() from SpectralModel so this syntax won't work now,
I updated the .from_dict() test and the example in analysis.py to fix this

cdeil commented:
@adonath - Merge?

cdeil commented:
I think all comments by @adonath and me are addressed, CI is green.
As discussed offline today, we'll iterate on this anyways in the next weeks.

Tomorrow I'll move this code to `gammapy.modeling` (see #2319), so merging this in now.
